Broker crashes if there is an empty log segment

**Describe the bug**

[logs](https://console.cloud.google.com/logs/viewer?expandAll=false&timestamp=2021-05-04T14%3A09%3A56.909000000Z&project=camunda-cloud-240911&minLogLevel=0&customFacets&limitCustomFacetWidth=true&interval=JUMP_TO_TIME&scrollTimestamp=2021-05-04T09%3A42%3A12.919247000Z&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%22de990736-75b6-4e2b-8b51-65b2ccd9174d-zeebe%22%0Aresource.labels.cluster_name%3D%22ultrachaos%22%0Aresource.labels.pod_name%3D%22zeebe-0%22%0Aresource.labels.project_id%3D%22camunda-cloud-240911%22%0Aresource.labels.location%3D%22europe-west1-d%22%0Aresource.labels.container_name%3D%22zeebe%22%0Atimestamp%3D%222021-05-04T09%3A42%3A10.873656000Z%22%0AinsertId%3D%2295xz4qg4ckwrmf%22)

```
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[atomix-utils-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.nio.Buffer.createPositionException(Unknown Source) ~[?:?]
	at java.nio.Buffer.position(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.position(Unknown Source) ~[?:?]
	at java.nio.MappedByteBuffer.position(Unknown Source) ~[?:?]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.reset(MappedJournalSegmentWriter.java:202) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.<init>(MappedJournalSegmentWriter.java:66) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.createWriter(JournalSegment.java:151) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.<init>(JournalSegment.java:62) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.newSegment(SegmentedJournal.java:461) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegment(SegmentedJournal.java:472) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegments(SegmentedJournal.java:520) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:228) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:85) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:148) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.log.RaftLogBuilder.build(RaftLogBuilder.java:149) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:233) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:155) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:257) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:231) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:175) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:145) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:103) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:113) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:201) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
	at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:203) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.lambda$start$1(DefaultPartitionService.java:80) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
	at java.util.concurrent.ConcurrentHashMap$ValueSpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.start(DefaultPartitionService.java:81) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.core.Atomix.lambda$startServices$0(Atomix.java:178) ~[atomix-1.0.0-alpha7.jar:1.0.0-alpha7]
```
I inspected disk of the broker.  The segment that caused the problem had a file length of 0. So even the `JournalSegmentDescriptor` was not written to the file.
However what is more strange was that the file was "raft-partition-partition-5-0.log". The id of the segment is 0 which should never happen. The segment id's start from 1 always.

**Expected behavior**

* There should not be any segment with id 0.
* If it is the last segment, then it is most likely empty because the broker crashed before writing anything to the file. Hence it would be safe to ignore that segment and create a new one.

**TLDR**: a partial write caused a zero-length file to be loaded into a buffer which was then used with the SBE decoder to read the fields. Since the buffer was empty, all of the fields were read as zero. This is why there was a file 'raft-partition-partition-5-0.log' even though no segment should have a 0 id and why we failed while reading from an empty buffer.

Long version:
Looking into the data folder a few weird things stand out. First, file 'raft-partition-partition-5-0.log' has an id of zero which should never happen since the initial id is 1 and they're always incremented. Second, the same file has a file size of zero:

```
root@zeebe-0:/usr/local/zeebe/data/raft-partition/partitions/5# ls -lh
total 257M
drwxr-xr-x 2 root root 4.0K May  4 09:42 pending
-rw-r--r-- 1 root root    0 May  4 09:42 raft-partition-partition-5-0.log
-rw-r--r-- 1 root root 128M May  4 09:33 raft-partition-partition-5-368.log
-rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-369.log
-rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-370.log
-rw-r--r-- 1 root root   78 May  3 13:00 raft-partition-partition-5.conf
-rw-r--r-- 1 root root  256 May  7 14:33 raft-partition-partition-5.meta
drwxr-xr-x 2 root root  16K May  4 09:35 runtime
drwxr-xr-x 3 root root 4.0K May  4 09:31 snapshots
```

At first, the empty file pointed to a partial write and the zero id pointed to corruption of a previous file. However, the key thing to note is that the last log segment 'raft-partition-partition-5-370.log' may seem to have the full 128MB but its actual allocated size is 0:

```
root@zeebe-0:/usr/local/zeebe/data/raft-partition/partitions/5# ls -lhs 
total 257M
4.0K drwxr-xr-x 2 root root 4.0K May  4 09:42 pending
   0 -rw-r--r-- 1 root root    0 May  4 09:42 raft-partition-partition-5-0.log
128M -rw-r--r-- 1 root root 128M May  4 09:33 raft-partition-partition-5-368.log
128M -rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-369.log
   0 -rw-r--r-- 1 root root 128M May  4 09:36 raft-partition-partition-5-370.log
4.0K -rw-r--r-- 1 root root   78 May  3 13:00 raft-partition-partition-5.conf
4.0K -rw-r--r-- 1 root root  256 May  7 14:49 raft-partition-partition-5.meta
 16K drwxr-xr-x 2 root root  16K May  4 09:35 runtime
4.0K drwxr-xr-x 3 root root 4.0K May  4 09:31 snapshots
```

So the partial write actually occurred in 'raft-partition-partition-5-370.log'. The timeline of events is the following:

On 04/05/2021 - 11:36, zeebe-0 crashed while creating a log segment with id 370. This is supported by the fact that the last log before the restart is at 11:35:58 CEST and the time of creation of the file is 09:36:51 UTC. It is also supported by the fact that its real size is 0:
```
root@zeebe-0:/usr/local/zeebe/data/raft-partition/partitions/5# ls -lhs --time-style=full-iso
total 257M
4.0K drwxr-xr-x 2 root root 4.0K 2021-05-04 09:42:08.182905902 +0000 pending
   0 -rw-r--r-- 1 root root    0 2021-05-04 09:42:10.871392216 +0000 raft-partition-partition-5-0.log
128M -rw-r--r-- 1 root root 128M 2021-05-04 09:33:55.268323034 +0000 raft-partition-partition-5-368.log
128M -rw-r--r-- 1 root root 128M 2021-05-04 09:36:51.043576425 +0000 raft-partition-partition-5-369.log
   0 -rw-r--r-- 1 root root 128M 2021-05-04 09:36:51.051576891 +0000 raft-partition-partition-5-370.log
4.0K -rw-r--r-- 1 root root   78 2021-05-03 13:00:56.074183086 +0000 raft-partition-partition-5.conf
4.0K -rw-r--r-- 1 root root  256 2021-05-07 14:49:21.696284442 +0000 raft-partition-partition-5.meta
 16K drwxr-xr-x 2 root root  16K 2021-05-04 09:35:53.141198840 +0000 runtime
4.0K drwxr-xr-x 3 root root 4.0K 2021-05-04 09:31:52.532163536 +0000 snapshots
```

On 04/05/2021 - 11:39, broker zeebe-0 restarts. When loading segment 'raft-partition-partition-5-370.log', we allocate a buffer and read the empty file into it.
https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java#L507
However, we don't check if any bytes were actually read so at this point we happily proceed with an empty buffer. Then we use it to create the segment descriptor object and, since we don't check that the SBE schema id, the template id and schema version match, we don't detect that this buffer is empty and the descriptor is successfully created.

https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptor.java#L68

Then we use the in-memory descriptor to load it again (I'm not sure why, might be a left over from before). However, since the descriptor was made by wrapping an empty buffer, when reading its fields we can only read 0. So when we create the segment file in `loadSegment`, we create it with the id 0. 

https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java#L463

When the creating the actual JournalSegment object, we map the segment file into memory. However, the length of the mapped region is given also by the descriptor so this is also set to 0.
https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegment.java#L59

Lastly, we create the segment writer which calls reset to advance its buffer's position past the descriptor bytes.
https://github.com/camunda-cloud/zeebe/blob/f5973af369d729387240278533bce251d4899827/journal/src/main/java/io/camunda/zeebe/journal/file/MappedJournalSegmentWriter.java#L199

Since we used length 0 when mapping the segment file, the limit of the mapped buffer is zero and the call to `position` fails because we've passed a position larger than the limit:
```
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[atomix-utils-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.lang.IllegalArgumentException: newPosition > limit: (9 > 0)
	at java.nio.Buffer.createPositionException(Unknown Source) ~[?:?]
	at java.nio.Buffer.position(Unknown Source) ~[?:?]
	at java.nio.ByteBuffer.position(Unknown Source) ~[?:?]
	at java.nio.MappedByteBuffer.position(Unknown Source) ~[?:?]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.reset(MappedJournalSegmentWriter.java:202) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.<init>(MappedJournalSegmentWriter.java:66) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.createWriter(JournalSegment.java:151) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.JournalSegment.<init>(JournalSegment.java:62) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.newSegment(SegmentedJournal.java:461) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegment(SegmentedJournal.java:472) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.loadSegments(SegmentedJournal.java:520) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:228) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:85) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:148) ~[zeebe-journal-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.log.RaftLogBuilder.build(RaftLogBuilder.java:149) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:233) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:155) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:257) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:231) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:175) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:145) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:103) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:113) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:201) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
	at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:203) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.lambda$start$1(DefaultPartitionService.java:80) ~[atomix-cluster-1.0.0-alpha7.jar:1.0.0-alpha7]
        ....
```

There are several checks that we can add (I've tried to point them out in the previous paragraphs) that would've made this issue easier to diagnose. However, they wouldn't have resolve the issue because the root cause is a broker crash causing a partially written log segment. This raises the same issue that we've encountered before, namely, that we need to be able to distinguish partial writes that can be ignored/deleted from actual corruption that we must preserved and dealt with manually to avoid losing acknowledged data. 

One solution can be writing a zero byte as the first byte of the descriptor (currently we just write a version 1 as a placeholder). Then we can write the actual descriptor, flush it and then write the descriptor version as the starting magic byte. This way, when reading from the file, we can tell if was fully written by looking at the magic byte. The good news is that, since we already write this byte, we can do this in a backwards compatible way after the release (good thing too since it's 17:59 on the Friday before the release)

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptor.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptor.java
@@ -69,7 +69,13 @@ public final class JournalSegmentDescriptor {
 
     if (headerDecoder.schemaId() != segmentDescriptorDecoder.sbeSchemaId()
         || headerDecoder.templateId() != segmentDescriptorDecoder.sbeTemplateId()) {
-      throw new CorruptedLogException("Cannot read segment descriptor. Header does not match.");
+      throw new CorruptedLogException(
+          String.format(
+              "Cannot read segment descriptor. Read schema and template ids ('%d' and '%d') don't match expected '%d' and %d'.",
+              headerDecoder.schemaId(),
+              headerDecoder.templateId(),
+              segmentDescriptorDecoder.sbeSchemaId(),
+              segmentDescriptorDecoder.sbeTemplateId()));
     }
 
     segmentDescriptorDecoder.wrap(

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptor.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptor.java
@@ -18,6 +18,7 @@ package io.camunda.zeebe.journal.file;
 
 import static com.google.common.base.Preconditions.checkArgument;
 
+import io.camunda.zeebe.journal.file.record.CorruptedLogException;
 import java.nio.ByteBuffer;
 import java.util.Objects;
 import org.agrona.MutableDirectBuffer;
@@ -52,11 +53,11 @@ public final class JournalSegmentDescriptor {
   private final int encodedLength;
 
   private final SegmentDescriptorDecoder segmentDescriptorDecoder = new SegmentDescriptorDecoder();
-  private final MessageHeaderDecoder messageHeaderDecoder = new MessageHeaderDecoder();
+  private final MessageHeaderDecoder headerDecoder = new MessageHeaderDecoder();
   private final MutableDirectBuffer directBuffer = new UnsafeBuffer();
 
   private final SegmentDescriptorEncoder segmentDescriptorEncoder = new SegmentDescriptorEncoder();
-  private final MessageHeaderEncoder messageHeaderEncoder = new MessageHeaderEncoder();
+  private final MessageHeaderEncoder headerEncoder = new MessageHeaderEncoder();
 
   public JournalSegmentDescriptor(final ByteBuffer buffer) {
     directBuffer.wrap(buffer);
@@ -64,20 +65,24 @@ public final class JournalSegmentDescriptor {
     /* The first byte of the buffer contains the version at which the descriptor is written.
      Currently we have only one version, so we can ignore it. We can use this version in future
     when needed. Note that we can add fields to SBE schema of the descriptor without changing this version. */
-    messageHeaderDecoder.wrap(directBuffer, VERSION_LENGTH);
+    headerDecoder.wrap(directBuffer, VERSION_LENGTH);
+
+    if (headerDecoder.schemaId() != segmentDescriptorDecoder.sbeSchemaId()
+        || headerDecoder.templateId() != segmentDescriptorDecoder.sbeTemplateId()) {
+      throw new CorruptedLogException("Cannot read segment descriptor. Header does not match.");
+    }
+
     segmentDescriptorDecoder.wrap(
         directBuffer,
-        VERSION_LENGTH + messageHeaderDecoder.encodedLength(),
-        messageHeaderDecoder.blockLength(),
-        messageHeaderDecoder.version());
+        VERSION_LENGTH + headerDecoder.encodedLength(),
+        headerDecoder.blockLength(),
+        headerDecoder.version());
 
     id = segmentDescriptorDecoder.id();
     index = segmentDescriptorDecoder.index();
     maxSegmentSize = segmentDescriptorDecoder.maxSegmentSize();
     encodedLength =
-        VERSION_LENGTH
-            + messageHeaderDecoder.encodedLength()
-            + segmentDescriptorDecoder.encodedLength();
+        VERSION_LENGTH + headerDecoder.encodedLength() + segmentDescriptorDecoder.encodedLength();
   }
 
   private JournalSegmentDescriptor(final long id, final long index, final int maxSegmentSize) {
@@ -158,7 +163,7 @@ public final class JournalSegmentDescriptor {
     directBuffer.wrap(buffer);
     directBuffer.putByte(0, VERSION);
 
-    messageHeaderEncoder
+    headerEncoder
         .wrap(directBuffer, VERSION_LENGTH)
         .blockLength(segmentDescriptorEncoder.sbeBlockLength())
         .templateId(segmentDescriptorEncoder.sbeTemplateId())
@@ -166,7 +171,7 @@ public final class JournalSegmentDescriptor {
         .version(segmentDescriptorEncoder.sbeSchemaVersion());
 
     segmentDescriptorEncoder
-        .wrap(directBuffer, VERSION_LENGTH + messageHeaderEncoder.encodedLength())
+        .wrap(directBuffer, VERSION_LENGTH + headerEncoder.encodedLength())
         .id(id)
         .index(index)
         .maxSegmentSize(maxSegmentSize);

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentFile.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegmentFile.java
@@ -30,6 +30,7 @@ import java.nio.file.StandardOpenOption;
  * @author <a href="http://github.com/kuujo">Jordan Halterman</a>
  */
 final class JournalSegmentFile {
+
   private static final char PART_SEPARATOR = '-';
   private static final char EXTENSION_SEPARATOR = '.';
   private static final String EXTENSION = "log";
@@ -62,23 +63,32 @@ final class JournalSegmentFile {
     checkNotNull(journalName, "journalName cannot be null");
     checkNotNull(fileName, "fileName cannot be null");
 
-    final int partSeparator = fileName.lastIndexOf(PART_SEPARATOR);
-    final int extensionSeparator = fileName.lastIndexOf(EXTENSION_SEPARATOR);
+    if (getSegmentIdFromPath(fileName) == -1) {
+      return false;
+    }
+
+    return fileName.startsWith(journalName);
+  }
+
+  /** Returns the segment's id or -1 if the log segment name was not correctly formatted. */
+  static int getSegmentIdFromPath(final String name) {
+    checkNotNull(name, "name cannot be null");
+
+    final int partSeparator = name.lastIndexOf(PART_SEPARATOR);
+    final int extensionSeparator = name.lastIndexOf(EXTENSION_SEPARATOR);
 
     if (extensionSeparator == -1
         || partSeparator == -1
         || extensionSeparator < partSeparator
-        || !fileName.endsWith(EXTENSION)) {
-      return false;
+        || !name.endsWith(EXTENSION)) {
+      return -1;
     }
 
-    for (int i = partSeparator + 1; i < extensionSeparator; i++) {
-      if (!Character.isDigit(fileName.charAt(i))) {
-        return false;
-      }
+    try {
+      return Integer.parseInt(name.substring(partSeparator + 1, extensionSeparator));
+    } catch (final NumberFormatException e) {
+      return -1;
     }
-
-    return fileName.startsWith(journalName);
   }
 
   /** Creates a segment file for the given directory, log name, segment ID, and segment version. */

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
@@ -503,17 +503,14 @@ public class SegmentedJournal implements Journal {
       final File file = files.get(i);
 
       try {
+        log.debug("Found segment file: {}", file.getName());
         final JournalSegmentDescriptor descriptor = readDescriptor(file);
         final JournalSegment segment = loadSegment(descriptor.id());
 
-        if (i > 0 && segments.get(i - 1).lastIndex() + 1 != segment.index()) {
-          throw new CorruptedLogException(
-              String.format(
-                  "Log segment %s is not aligned with previous segment %s.",
-                  segments.get(i - 1), segment));
+        if (i > 0) {
+          checkForIndexGaps(segments.get(i - 1), segment);
         }
 
-        log.debug("Found segment: {} ({})", descriptor.id(), file.getName());
         segments.add(segment);
       } catch (final CorruptedLogException e) {
         if (handleSegmentCorruption(files, segments, i)) {
@@ -525,6 +522,15 @@ public class SegmentedJournal implements Journal {
     }
 
     return segments;
+  }
+
+  private void checkForIndexGaps(final JournalSegment prevSegment, final JournalSegment segment) {
+    if (prevSegment.lastIndex() != segment.index() - 1) {
+      throw new CorruptedLogException(
+          String.format(
+              "Log segment %s is not aligned with previous segment %s (last index: %d).",
+              segment, prevSegment, prevSegment.lastIndex()));
+    }
   }
 
   /** Returns true if segments after corrupted segment were deleted; false, otherwise */
@@ -542,15 +548,20 @@ public class SegmentedJournal implements Journal {
     }
 
     log.debug(
-        "Found corrupted segment after last ack'ed index {}. Deleting it and subsequent segments",
-        lastWrittenIndex);
+        "Found corrupted segment after last ack'ed index {}. Deleting segments {} - {}",
+        lastWrittenIndex,
+        files.get(failedIndex).getName(),
+        files.get(files.size() - 1).getName());
 
     for (int i = failedIndex; i < files.size(); i++) {
       final File file = files.get(i);
       try {
         Files.delete(file.toPath());
       } catch (IOException e) {
-        log.warn("Failed to delete log segment {}", file.getName(), e);
+        throw new JournalException(
+            String.format(
+                "Failed to delete log segment '%s' when handling corruption.", file.getName()),
+            e);
       }
     }
 

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
@@ -24,6 +24,7 @@ import io.camunda.zeebe.journal.Journal;
 import io.camunda.zeebe.journal.JournalException;
 import io.camunda.zeebe.journal.JournalReader;
 import io.camunda.zeebe.journal.JournalRecord;
+import io.camunda.zeebe.journal.file.record.CorruptedLogException;
 import java.io.File;
 import java.io.IOException;
 import java.io.RandomAccessFile;
@@ -502,15 +503,7 @@ public class SegmentedJournal implements Journal {
       // If the file looks like a segment file, attempt to load the segment.
       if (JournalSegmentFile.isSegmentFile(name, file)) {
         final JournalSegmentFile segmentFile = new JournalSegmentFile(file);
-        final ByteBuffer buffer = ByteBuffer.allocate(JournalSegmentDescriptor.getEncodingLength());
-        try (final FileChannel channel = openChannel(file)) {
-          channel.read(buffer);
-          buffer.flip();
-        } catch (final IOException e) {
-          throw new JournalException(e);
-        }
-
-        final JournalSegmentDescriptor descriptor = new JournalSegmentDescriptor(buffer);
+        final JournalSegmentDescriptor descriptor = readDescriptor(file);
 
         // Load the segment.
         final JournalSegment segment = loadSegment(descriptor.id());
@@ -544,6 +537,33 @@ public class SegmentedJournal implements Journal {
     }
 
     return segments.values();
+  }
+
+  private JournalSegmentDescriptor readDescriptor(final File file) {
+    final int descriptorLength = JournalSegmentDescriptor.getEncodingLength();
+    if (file.length() < descriptorLength) {
+      throw new CorruptedLogException(
+          String.format(
+              "Log segment is smaller than a segment descriptor (%d < %d).",
+              file.length(), descriptorLength));
+    }
+
+    final ByteBuffer buffer = ByteBuffer.allocate(descriptorLength);
+    try (final FileChannel channel = openChannel(file)) {
+      final int readBytes = channel.read(buffer);
+
+      if (readBytes != -1 && readBytes < descriptorLength) {
+        throw new JournalException(
+            String.format(
+                "Expected to read segment descriptor (%d bytes) but only read %d bytes.",
+                descriptorLength, readBytes));
+      }
+      buffer.flip();
+    } catch (final IOException e) {
+      throw new JournalException(e);
+    }
+
+    return new JournalSegmentDescriptor(buffer);
   }
 
   public void closeReader(final SegmentedJournalReader segmentedJournalReader) {

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
@@ -30,14 +30,16 @@ import java.io.IOException;
 import java.io.RandomAccessFile;
 import java.nio.ByteBuffer;
 import java.nio.channels.FileChannel;
+import java.nio.file.Files;
 import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
-import java.util.Iterator;
+import java.util.Comparator;
+import java.util.List;
 import java.util.Map;
-import java.util.Map.Entry;
 import java.util.NavigableMap;
 import java.util.SortedMap;
-import java.util.TreeMap;
 import java.util.concurrent.ConcurrentSkipListMap;
 import java.util.concurrent.locks.StampedLock;
 import org.agrona.DirectBuffer;
@@ -494,49 +496,83 @@ public class SegmentedJournal implements Journal {
   protected Collection<JournalSegment> loadSegments() {
     // Ensure log directories are created.
     directory.mkdirs();
+    final List<JournalSegment> segments = new ArrayList<>();
 
-    final TreeMap<Long, JournalSegment> segments = new TreeMap<>();
+    final List<File> files = getSortedLogSegments();
+    for (int i = 0; i < files.size(); i++) {
+      final File file = files.get(i);
 
-    // Iterate through all files in the log directory.
-    for (final File file : directory.listFiles(File::isFile)) {
-
-      // If the file looks like a segment file, attempt to load the segment.
-      if (JournalSegmentFile.isSegmentFile(name, file)) {
-        final JournalSegmentFile segmentFile = new JournalSegmentFile(file);
+      try {
         final JournalSegmentDescriptor descriptor = readDescriptor(file);
-
-        // Load the segment.
         final JournalSegment segment = loadSegment(descriptor.id());
 
-        // Add the segment to the segments list.
-        log.debug(
-            "Found segment: {} ({})", segment.descriptor().id(), segmentFile.file().getName());
-        segments.put(segment.index(), segment);
+        if (i > 0 && segments.get(i - 1).lastIndex() + 1 != segment.index()) {
+          throw new CorruptedLogException(
+              String.format(
+                  "Log segment %s is not aligned with previous segment %s.",
+                  segments.get(i - 1), segment));
+        }
+
+        log.debug("Found segment: {} ({})", descriptor.id(), file.getName());
+        segments.add(segment);
+      } catch (final CorruptedLogException e) {
+        if (handleSegmentCorruption(files, segments, i)) {
+          return segments;
+        }
+
+        throw e;
       }
     }
 
-    // Verify that all the segments in the log align with one another.
-    JournalSegment previousSegment = null;
-    boolean corrupted = false;
-    final Iterator<Entry<Long, JournalSegment>> iterator = segments.entrySet().iterator();
-    while (iterator.hasNext()) {
-      final JournalSegment segment = iterator.next().getValue();
-      if (previousSegment != null && previousSegment.lastIndex() != segment.index() - 1) {
-        log.warn(
-            "Journal is inconsistent. {} is not aligned with prior segment {}",
-            segment.file().file(),
-            previousSegment.file().file());
-        corrupted = true;
-      }
-      if (corrupted) {
-        segment.close();
-        segment.delete();
-        iterator.remove();
-      }
-      previousSegment = segment;
+    return segments;
+  }
+
+  /** Returns true if segments after corrupted segment were deleted; false, otherwise */
+  private boolean handleSegmentCorruption(
+      final List<File> files, final List<JournalSegment> segments, final int failedIndex) {
+    long lastSegmentIndex = 0;
+
+    if (!segments.isEmpty()) {
+      final JournalSegment previousSegment = segments.get(segments.size() - 1);
+      lastSegmentIndex = previousSegment.lastIndex();
     }
 
-    return segments.values();
+    if (lastWrittenIndex > lastSegmentIndex) {
+      return false;
+    }
+
+    log.debug(
+        "Found corrupted segment after last ack'ed index {}. Deleting it and subsequent segments",
+        lastWrittenIndex);
+
+    for (int i = failedIndex; i < files.size(); i++) {
+      final File file = files.get(i);
+      try {
+        Files.delete(file.toPath());
+      } catch (IOException e) {
+        log.warn("Failed to delete log segment {}", file.getName(), e);
+      }
+    }
+
+    return true;
+  }
+
+  /** Returns an array of valid log segments sorted by their id which may be empty but not null. */
+  private List<File> getSortedLogSegments() {
+    final File[] files =
+        directory.listFiles(file -> file.isFile() && JournalSegmentFile.isSegmentFile(name, file));
+
+    if (files == null) {
+      throw new IllegalStateException(
+          String.format(
+              "Could not list files in directory '%s'. Either the path doesn't point to a directory or an I/O error occurred.",
+              directory));
+    }
+
+    Arrays.sort(
+        files, Comparator.comparingInt(f -> JournalSegmentFile.getSegmentIdFromPath(f.getName())));
+
+    return Arrays.asList(files);
   }
 
   private JournalSegmentDescriptor readDescriptor(final File file) {

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptorTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptorTest.java
@@ -21,10 +21,10 @@ import static org.assertj.core.api.Assertions.assertThat;
 import java.nio.ByteBuffer;
 import org.junit.jupiter.api.Test;
 
-public class JournalSegmentDescriptorTest {
+class JournalSegmentDescriptorTest {
 
   @Test
-  public void shouldWriteAndReadDescriptor() {
+  void shouldWriteAndReadDescriptor() {
     // given
     JournalSegmentDescriptor descriptor =
         JournalSegmentDescriptor.builder()

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptorTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/JournalSegmentDescriptorTest.java
@@ -17,7 +17,9 @@
 package io.camunda.zeebe.journal.file;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
+import io.camunda.zeebe.journal.file.record.CorruptedLogException;
 import java.nio.ByteBuffer;
 import org.junit.jupiter.api.Test;
 
@@ -44,5 +46,15 @@ class JournalSegmentDescriptorTest {
     assertThat(descriptorRead.index()).isEqualTo(100);
     assertThat(descriptorRead.maxSegmentSize()).isEqualTo(1024);
     assertThat(descriptorRead.length()).isEqualTo(JournalSegmentDescriptor.getEncodingLength());
+  }
+
+  @Test
+  void shouldValidateDescriptorHeader() {
+    // given
+    final ByteBuffer buffer = ByteBuffer.allocate(JournalSegmentDescriptor.getEncodingLength());
+
+    // when/then
+    assertThatThrownBy(() -> new JournalSegmentDescriptor(buffer))
+        .isInstanceOf(CorruptedLogException.class);
   }
 }

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/JournalTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/JournalTest.java
@@ -606,7 +606,7 @@ class JournalTest {
     assertThat(reader.next()).isEqualTo(lastRecord);
   }
 
-  private PersistedJournalRecord copyRecord(final JournalRecord record) {
+  static PersistedJournalRecord copyRecord(final JournalRecord record) {
     final DirectBuffer data = record.data();
     final byte[] buffer = new byte[data.capacity()];
     data.getBytes(0, buffer);

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/JournalTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/JournalTest.java
@@ -32,6 +32,7 @@ import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 import java.util.function.Consumer;
 import org.agrona.DirectBuffer;
 import org.agrona.concurrent.UnsafeBuffer;
@@ -112,7 +113,7 @@ class JournalTest {
   }
 
   @Test
-  public void shouldReadMultipleRecord() {
+  void shouldReadMultipleRecord() {
     // given
     final var firstRecord = journal.append(1, data);
     final var secondRecord = journal.append(20, dataOther);
@@ -128,7 +129,7 @@ class JournalTest {
   }
 
   @Test
-  public void shouldAppendAndReadMultipleRecordsInOrder() {
+  void shouldAppendAndReadMultipleRecordsInOrder() {
     // when
     for (int i = 0; i < 10; i++) {
       final var recordAppended = journal.append(i + 10, data);
@@ -185,7 +186,7 @@ class JournalTest {
     // then
     assertThat(journal.isEmpty()).isTrue();
     assertThat(journal.getLastIndex()).isEqualTo(1);
-    final var record = journal.append(asqn++, data);
+    final var record = journal.append(asqn, data);
     assertThat(record.index()).isEqualTo(2);
   }
 
@@ -205,7 +206,7 @@ class JournalTest {
 
     // then
     assertThat(journal.getLastIndex()).isEqualTo(1);
-    final var record = journal.append(asqn++, data);
+    final var record = journal.append(asqn, data);
     assertThat(record.index()).isEqualTo(2);
 
     // then
@@ -541,7 +542,7 @@ class JournalTest {
   }
 
   @Test
-  public void shouldInvalidateAllEntries() throws Exception {
+  void shouldInvalidateAllEntries() throws Exception {
     // given
     data.wrap("000".getBytes(StandardCharsets.UTF_8));
     final var firstRecord = copyRecord(journal.append(data));
@@ -565,14 +566,13 @@ class JournalTest {
   }
 
   @Test
-  public void shouldDetectCorruptedEntry() throws Exception {
+  void shouldDetectCorruptedEntry() throws Exception {
     // given
     data.wrap("000".getBytes(StandardCharsets.UTF_8));
     journal.append(data);
     final var secondRecord = copyRecord(journal.append(data));
-    assertThat(directory.toFile().listFiles()).hasSize(1);
-    assertThat(directory.toFile().listFiles()[0].listFiles()).hasSize(1);
-    final File log = directory.toFile().listFiles()[0].listFiles()[0];
+    final File dataFile = Objects.requireNonNull(directory.toFile().listFiles())[0];
+    final File log = Objects.requireNonNull(dataFile.listFiles())[0];
 
     // when
     journal.close();
@@ -585,14 +585,13 @@ class JournalTest {
   }
 
   @Test
-  public void shouldDeletePartiallyWrittenEntry() throws Exception {
+  void shouldDeletePartiallyWrittenEntry() throws Exception {
     // given
     data.wrap("000".getBytes(StandardCharsets.UTF_8));
     final var firstRecord = copyRecord(journal.append(data));
     final var secondRecord = copyRecord(journal.append(data));
-    assertThat(directory.toFile().listFiles()).hasSize(1);
-    assertThat(directory.toFile().listFiles()[0].listFiles()).hasSize(1);
-    final File log = directory.toFile().listFiles()[0].listFiles()[0];
+    final File dataFile = Objects.requireNonNull(directory.toFile().listFiles())[0];
+    final File log = Objects.requireNonNull(dataFile.listFiles())[0];
 
     // when
     journal.close();

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/LogCorrupter.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/LogCorrupter.java
@@ -19,8 +19,6 @@ import java.nio.ByteBuffer;
 
 public class LogCorrupter {
 
-  private static final int BUFFER_SIZE = 8 * 1024;
-
   /**
    * Corrupts the record associated with the specified index, if it's present in the file.
    *
@@ -29,12 +27,16 @@ public class LogCorrupter {
    * @return true if the specified record was successfully corrupted; otherwise, returns false
    */
   public static boolean corruptRecord(final File file, final long index) throws IOException {
-    final byte[] bytes = new byte[BUFFER_SIZE];
-    int read = 0;
+    final byte[] bytes = new byte[Math.toIntExact(file.length())];
+    int read;
 
     try (final BufferedInputStream in = new BufferedInputStream(new FileInputStream(file))) {
-      while (in.available() > 0 && read < bytes.length) {
-        read += in.read(bytes, read, Math.min(bytes.length, in.available()) - read);
+      final int available = in.available();
+      read = in.read(bytes, 0, bytes.length);
+
+      // reached EOF
+      if (read == -1) {
+        read = available;
       }
     }
 
@@ -47,6 +49,28 @@ public class LogCorrupter {
     }
 
     return true;
+  }
+
+  public static void corruptDescriptor(final File file) throws IOException {
+    final byte[] bytes = new byte[JournalSegmentDescriptor.getEncodingLength()];
+    int read;
+
+    try (final BufferedInputStream in = new BufferedInputStream(new FileInputStream(file))) {
+      final int available = in.available();
+      read = in.read(bytes, 0, bytes.length);
+
+      // reached EOF
+      if (read == -1) {
+        read = available;
+      }
+    }
+
+    final byte schemaId = bytes[MessageHeaderDecoder.schemaIdEncodingOffset() + 1];
+    bytes[MessageHeaderDecoder.schemaIdEncodingOffset() + 1] = (byte) ~schemaId;
+
+    try (final BufferedOutputStream out = new BufferedOutputStream(new FileOutputStream(file))) {
+      out.write(bytes, 0, read);
+    }
   }
 
   private static boolean corruptRecord(final byte[] bytes, final long targetIndex) {

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
@@ -16,7 +16,7 @@
 package io.camunda.zeebe.journal.file;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
 
 import io.camunda.zeebe.journal.JournalReader;
 import io.camunda.zeebe.journal.JournalRecord;
@@ -27,6 +27,7 @@ import java.io.File;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
+import java.util.Objects;
 import org.agrona.DirectBuffer;
 import org.agrona.concurrent.UnsafeBuffer;
 import org.junit.jupiter.api.Test;
@@ -365,15 +366,69 @@ class SegmentedJournalTest {
   }
 
   @Test
-  void shouldDetectPartiallyWrittenDescriptor() throws Exception {
+  void shouldHandlePartiallyWrittenDescriptor() throws Exception {
     // given
-    final File data = directory.resolve("data").toFile();
-    assertThat(data.mkdirs()).isTrue();
-    final File emptyLog = new File(data, "journal-1.log");
+    final File dataFile = directory.resolve("data").toFile();
+    assertThat(dataFile.mkdirs()).isTrue();
+    final File emptyLog = new File(dataFile, "journal-1.log");
     assertThat(emptyLog.createNewFile()).isTrue();
 
+    // when
+    final var journal = openJournal(10);
+    final var reader = journal.openReader();
+    final var record = journal.append(data);
+
+    // then
+    assertThat(journal.getFirstIndex()).isEqualTo(record.index());
+    assertThat(journal.getLastIndex()).isEqualTo(record.index());
+    assertThat(reader.next()).isEqualTo(record);
+    assertThat(reader.hasNext()).isFalse();
+  }
+
+  @Test
+  void shouldHandleCorruptionAtDescriptorWithoutAckedEntries() throws Exception {
+    // given
+    var journal = openJournal(1);
+    journal.close();
+    final File dataFile = directory.resolve("data").toFile();
+    final File logFile =
+        Objects.requireNonNull(dataFile.listFiles(f -> f.getName().endsWith(".log")))[0];
+    LogCorrupter.corruptDescriptor(logFile);
+
     // when/then
-    assertThatThrownBy(() -> openJournal(10)).isInstanceOf(CorruptedLogException.class);
+    journal = openJournal(1);
+    final var reader = journal.openReader();
+    final var record = journal.append(data);
+
+    // then
+    assertThat(journal.getFirstIndex()).isEqualTo(record.index());
+    assertThat(journal.getLastIndex()).isEqualTo(record.index());
+    assertThat(reader.next()).isEqualTo(record);
+    assertThat(reader.hasNext()).isFalse();
+  }
+
+  @Test
+  void shouldDetectCorruptionAtDescriptorWithAckedEntries() throws Exception {
+    // given
+    final var journal = openJournal(1);
+    final var record = JournalTest.copyRecord(journal.append(data));
+
+    journal.close();
+    final File dataFile = directory.resolve("data").toFile();
+    final File logFile =
+        Objects.requireNonNull(dataFile.listFiles(f -> f.getName().endsWith(".log")))[0];
+    LogCorrupter.corruptDescriptor(logFile);
+
+    // when/then
+    assertThatThrownBy(
+            () ->
+                SegmentedJournal.builder()
+                    .withDirectory(directory.resolve("data").toFile())
+                    .withMaxSegmentSize(entrySize + JournalSegmentDescriptor.getEncodingLength())
+                    .withJournalIndexDensity(journalIndexDensity)
+                    .withLastWrittenIndex(record.index())
+                    .build())
+        .isInstanceOf(CorruptedLogException.class);
   }
 
   private SegmentedJournal openJournal(final float entriesPerSegment) {

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
@@ -408,10 +408,36 @@ class SegmentedJournalTest {
   }
 
   @Test
+  void shouldHandleCorruptionAtDescriptorWithSomeAckedEntries() throws Exception {
+    // given
+    var journal = openJournal(1);
+    final var firstRecord = JournalTest.copyRecord(journal.append(data));
+    journal.append(data);
+
+    journal.close();
+    final File dataFile = directory.resolve("data").toFile();
+    final File logFile =
+        Objects.requireNonNull(dataFile.listFiles(f -> f.getName().endsWith("2.log")))[0];
+    LogCorrupter.corruptDescriptor(logFile);
+
+    // when/then
+    journal = openJournal(1);
+    final var reader = journal.openReader();
+    final var lastRecord = journal.append(data);
+
+    // then
+    assertThat(journal.getFirstIndex()).isEqualTo(firstRecord.index());
+    assertThat(journal.getLastIndex()).isEqualTo(lastRecord.index());
+    assertThat(reader.next()).isEqualTo(firstRecord);
+    assertThat(reader.next()).isEqualTo(lastRecord);
+    assertThat(reader.hasNext()).isFalse();
+  }
+
+  @Test
   void shouldDetectCorruptionAtDescriptorWithAckedEntries() throws Exception {
     // given
     final var journal = openJournal(1);
-    final var record = JournalTest.copyRecord(journal.append(data));
+    final long index = journal.append(data).index();
 
     journal.close();
     final File dataFile = directory.resolve("data").toFile();
@@ -426,7 +452,7 @@ class SegmentedJournalTest {
                     .withDirectory(directory.resolve("data").toFile())
                     .withMaxSegmentSize(entrySize + JournalSegmentDescriptor.getEncodingLength())
                     .withJournalIndexDensity(journalIndexDensity)
-                    .withLastWrittenIndex(record.index())
+                    .withLastWrittenIndex(index)
                     .build())
         .isInstanceOf(CorruptedLogException.class);
   }

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
@@ -16,11 +16,14 @@
 package io.camunda.zeebe.journal.file;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import io.camunda.zeebe.journal.JournalReader;
 import io.camunda.zeebe.journal.JournalRecord;
+import io.camunda.zeebe.journal.file.record.CorruptedLogException;
 import io.camunda.zeebe.journal.file.record.RecordData;
 import io.camunda.zeebe.journal.file.record.SBESerializer;
+import java.io.File;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
@@ -359,6 +362,18 @@ class SegmentedJournalTest {
         .isEqualTo(indexBeforeClose.lookup(firstIndexedPosition).position());
     assertThat(indexAfterRestart.lookup(secondIndexedPosition).position())
         .isEqualTo(indexBeforeClose.lookup(secondIndexedPosition).position());
+  }
+
+  @Test
+  void shouldDetectPartiallyWrittenDescriptor() throws Exception {
+    // given
+    final File data = directory.resolve("data").toFile();
+    assertThat(data.mkdirs()).isTrue();
+    final File emptyLog = new File(data, "journal-1.log");
+    assertThat(emptyLog.createNewFile()).isTrue();
+
+    // when/then
+    assertThatThrownBy(() -> openJournal(10)).isInstanceOf(CorruptedLogException.class);
   }
 
   private SegmentedJournal openJournal(final float entriesPerSegment) {

--- a/journal/src/test/java/io/camunda/zeebe/journal/file/SparseJournalIndexTest.java
+++ b/journal/src/test/java/io/camunda/zeebe/journal/file/SparseJournalIndexTest.java
@@ -23,10 +23,10 @@ import io.camunda.zeebe.journal.JournalRecord;
 import org.junit.jupiter.api.Test;
 
 /** Sparse journal index test. */
-public class SparseJournalIndexTest {
+class SparseJournalIndexTest {
 
   @Test
-  public void shouldNotFindIndexWhenNotReachedDensity() {
+  void shouldNotFindIndexWhenNotReachedDensity() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
 
@@ -42,7 +42,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldFindIndexWhenReachedDensity() {
+  void shouldFindIndexWhenReachedDensity() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
 
@@ -60,7 +60,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldFindLowerIndexWhenNotReachedDensity() {
+  void shouldFindLowerIndexWhenNotReachedDensity() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
     // index entries
@@ -82,7 +82,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldFindNextIndexWhenReachedDensity() {
+  void shouldFindNextIndexWhenReachedDensity() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
     // index entries
@@ -106,7 +106,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldTruncateIndex() {
+  void shouldTruncateIndex() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
     // index entries
@@ -134,7 +134,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldTruncateCompleteIndex() {
+  void shouldTruncateCompleteIndex() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
     // index entries
@@ -165,7 +165,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldNotCompactIndex() {
+  void shouldNotCompactIndex() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
     // index entries
@@ -190,7 +190,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldCompactIndex() {
+  void shouldCompactIndex() {
     // given - every 5 index is added
     final JournalIndex index = new SparseJournalIndex(5);
     // index entries
@@ -217,7 +217,7 @@ public class SparseJournalIndexTest {
   }
 
   @Test
-  public void shouldFindAsqnWithInBound() {
+  void shouldFindAsqnWithInBound() {
     // given - every 2nd index is added
     final JournalIndex index = new SparseJournalIndex(2);
 

